### ISSUE TITLE
Is it Xcode 8 compatible?

### ISSUE THREAD

**Issue by RafaelPlantard** [Jintin/Swimat#123]
Can I use it in Xcode 8?

**Comment by saagarjha** [comment]
Xcode 8 breaks plugin support; new "plugins" must use Xcode's Source Editor Extensions. Swimat will have to be updated to use this before it would be compatible.

**Comment by Jintin** [comment]
@saagarjha is right, I'll work on the update thing a few days after update to XCode 8.
Thank you all for sharing information.

**Comment by RafaelPlantard** [comment]
We will be waiting for your amazing "plugin"...

**Comment by LivioGama** [comment]
I need it so much !

**Comment by SeanRobinson159** [comment]
Now that I have to use Xcode 8, without Swimat, I feel like a caveman again.

**Comment by Jintin** [comment]
Hey man, I've update our "plugin" to "extension" with swift 2.3.
You can either checkout `extension` branch and export to your own app or download the app file.
After restart xcode, you'll see a `Swimat` button in editor menu.
https://github.com/Jintin/Swimat/releases/download/1.0/Swimat.zip

Still need improvement and feedback. If you like to test it, it would be help. ^^

**Comment by saagarjha** [comment]
Hmm, I've built the source and run the project, but I can't get Swimat to appear in the Editor menu. I know that this worked in previous betas…are there any installation steps that I have missed?

**Comment by Jintin** [comment]
For test you might need run the command 

``` bash
sudo /usr/libexec/xpccachectl
```

or you can archive it to mac app.

**Comment by saagarjha** [comment]
It looks like Xcode was just taking its time to load the plugin. Anyways, it seems to work great!

**Comment by LivioGama** [comment]
What am I supposed to do with the unzipped file ?

**Comment by Jintin** [comment]
It's a Mac App, you can place it anywhere, but you have to open it first.
Then restart Xcode, it should be appear.

**Comment by LivioGama** [comment]
This is what I did, and it didn't appear :(

**Comment by saagarjha** [comment]
Restart Xcode a couple of times and wait…at least, that's what worked for me.

**Comment by Jintin** [comment]
I retry again.
- run the command `sudo /usr/libexec/xpccachectl`
- restart Mac
- open Xcode
- open Swimat app
- check format action in [Editor] -> [Swimat] -> [Format]

I think currently the Source Editor Extension feature in Xcode 8 is not very stable. You may find that the command is sometimes not showing up, or not available(grayed out).

**Comment by LivioGama** [comment]
Before reading this message, I tried those steps in different order during half an hour and never got it work. Then I tried with the order described below and got still no results. Let me try to understand what is the difference between my environment and yours, guys.
-Is your Mac in english or other language ?
-Do you have Swimat old plugin installed or not ?
-Do you still have Xcode 7 installed in parallel or not ?

**Comment by Jintin** [comment]
-Is your Mac in english or other language ?
english
-Do you have Swimat old plugin installed or not ?
yes, in Xcode 7.
-Do you still have Xcode 7 installed in parallel or not ?
yes, Xcode 7 is the main. 

I'll try on other computer a few days later. 
Sorry about the trap you face.
There are few resources now.

**Comment by loveuqian** [comment]
still not working now
waiting for the new plugin

**Comment by saagarjha** [comment]
Have you tried compiling from source?

**Comment by LivioGama** [comment]
I tried to compile the source also, nothing to do, it won't work

**Comment by RafaelPlantard** [comment]
It's not working with me too...

**Comment by andyast** [comment]
Not trying to pile on, but not working for me either.
OSX 10.11.6
Xcode 8.9 (8A218a)
I've tried compiling from source and the app.  

If I can provide any more assistance I'd be glad to help.

**Comment by awclives** [comment]
love this plugin.  does the new version work with swift 3?

**Comment by learnwithgabbar** [comment]
Swift 3.0?

**Comment by Jintin** [comment]
Here is the build your own step.
- Run the command sudo /usr/libexec/xpccachectl
- Sign app & extension using your developer certificate
- Build & Run (run extension)
- Restart Xcode
  You should see action in [Editor] -> [Swimat] -> [Format]

**Comment by LivioGama** [comment]
Do you need to be a Mac Developer ?

**Comment by RafaelPlantard** [comment]
@Jintin unfortunately I did twice your last tutorial of build in my own step, but it's not working 👎 

But trying a little bit more I just build + run the extension scheme above Xcode twice. And then run the app scheme and restart the Xcode and it's now working.

**Comment by mittenimraum** [comment]
It just doesn't appear in Xcode Version 8.0 (8A218a). Doesn't matter if i use the precompiled app or compile the project on my own. I'm using macOS Sierra already.

**UPDATE**

If the extension doesn't appear, check System Preferences -> Extensions. I had to enable Swimat as an extension for the Xcode Source Editor. It works now.

<img width="483" alt="bildschirmfoto 2016-09-21 um 16 41 41" src="https://cloud.githubusercontent.com/assets/1453177/18715789/ce8a704a-801a-11e6-8f15-0b479fa48454.png">

**Comment by andyast** [comment]
The above worked for me once I updated to the new version of macOS

**Comment by LivioGama** [comment]
I was on macOS Sierra and it worked for me too to check the checkbox

**Comment by LivioGama** [comment]
Any idea how I can make it work on "save" action and "run" action

**Comment by ddrachau** [comment]
Just to be sure, I found it too at _EDITOR_ -> Swimat 
I was searching for it at EDIT (I know that the comments said Editor, but I was blind :-) ), maybe this is a help for someone

**Comment by LivioGama** [comment]
No i was looking for it in the right place. It started to work for me only after mittenraum manipulation, after I updated to macOS Sierra.

**Comment by AaronBratcher** [comment]
Unfortunately it isn't working for me. On Sierra and Xcode 8 release version.

**Comment by Jintin** [comment]
save or run association is not available now.
I'll keep find solution.

**Comment by Jintin** [comment]
Swift 3.0 is also support since we don't care the real syntax, Swimat just arrange the structure.
It has pros & cons, but you don't have to worry about the swift version now.

**Comment by LivioGama** [comment]
Do you know Xcode "behaviors" ? Will try to figure out if there is a way with this

**Comment by awclives** [comment]
i just got it working by downloading the extension app.  on mac os sierra, xcode 8, swift 3.  did not have to build myself.

i used mittenimraum's suggestion.

EDIT: correction, i got it to install, but it's not working correctly yet.

**Comment by AaronBratcher** [comment]
I got it available in the menu thanks to the suggestion of mittenimraum, however it seems to lock up Xcode for me. First I get a warning that it is still working and then Xcode becomes unresponsive.

**Comment by awclives** [comment]
@AaronBratcher i can confirm that it's not working for me correctly either, as you describe.  It works occasionally.  

It is certainly not yet usable.

**Comment by Jintin** [comment]
@AaronBratcher the slow issue is solved, please update and try again.

**Comment by Jintin** [comment]
I've post a question to stackoverflow you may interest.
http://stackoverflow.com/questions/39807022/xcode8-extension
Or if somebody find the right solution please help!!!
Thanks.

**Comment by awclives** [comment]
has anyone gotten this working well with swift 3 and xcode 8?

**Comment by saagarjha** [comment]
Yes, I use it every day. It's a bit finicky to install, but when it works it does its job.

**Comment by awclives** [comment]
great, thanks @saagarjha and @Jintin .  I just got it working myself.

**Comment by awclives** [comment]
Still crashing XCode on an especially large file . . . 1500 lines.  Otherwise working fine.

**Comment by saagarjha** [comment]
Do you have a crash log? I can't reproduce this. For me, it just freezes for a couple of seconds, but finishes formatting fine.

**Comment by awclives** [comment]
Sorry, imprecise language.  It does not crash XCode, but it is unusable on the large file.  It freezes for several seconds and never finishes formatting (the "cancel" dialog pops up).  Also, one time Xcode stopped responding, and I had to force quit.

**Comment by saagarjha** [comment]
Yes, it freezes for ~20 seconds for me, but eventually finishes. How long are you waiting?

**Comment by Jintin** [comment]
I notice large file will slow than small file.
But 20 seconds is not acceptable, we'll try on large file to improve.
Thanks for report.

**Comment by Jintin** [comment]
I add several test and improvement today, could you update to try if it's better than old version?
In my case normal file(~ 500 lines) is less 1s, large file(> 1000 lines) is about 5s.
and if you still have trouble please open a new issue. ^^
my mac info:

```
MacBook Pro (Retina, 15-inch, Late 2013)
2.3 GHz Intel Core i7
```

**Comment by RafaelPlantard** [comment]
@Jintin you're an amazing guy 🗡

**Comment by Jintin** [comment]
With latest commit, large file(> 1000 lines) format is less than 2s.
I think everybody should update and try.

**Comment by RafaelPlantard** [comment]
@Jintin I guess that I found a mismatch format when using RxSwift:
<img width="853" alt="screen shot 2016-10-11 at 13 38 46" src="https://cloud.githubusercontent.com/assets/5968478/19279527/20534616-8fb8-11e6-83f2-515784fb568f.png">

**Comment by Jintin** [comment]
@RafaelPlantard Thank you.
You can open a new issue, and paste the code with this guide.
https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code

**Comment by Jintin** [comment]
@saagarjha @awclives 
Would you please update the latest version of code to test if the performance is better than before?
The latest version parser itself is about 10X faster.
Most pain point is the slow `String.characters.count`.
Now the bottleneck is the IO if there are lots file to update.
So if seldom line need update it will finish right away.
Any performance relative issue please refer to here: https://github.com/Jintin/Swimat/issues/131

**Comment by vietstone-ng** [comment]
Dont' know what is the issue with my environment that I can not get this beautiful extension for Xcode 8?
I even downloaded the release zip to try instead of archive the app myself. 
May be I'm using both Xcode 7, and Xcode 8 is the reason?
Haizz, nightmare come to me.

**Comment by saagarjha** [comment]
@vietstone You'll need Xcode 8 both to compile and to install Swimat. You may also need to restart them after installing.

**Comment by RafaelPlantard** [comment]
@Jintin I create another issue referencing that formatting problem #133

**Comment by Jintin** [comment]
@vietstone you can use `brew cask install swimat` to install, but you need to use it in xcode8.
Also make sure [System Preferences] -> [Extensions] -> [Xcode Source Editor] -> [Swimat] is checked. 
![](https://raw.githubusercontent.com/Jintin/Swimat/extension/README/setting.png)

**Comment by RafaelPlantard** [comment]
@Jintin this brew cask version is updated to latest version?

**Comment by Jintin** [comment]
@RafaelPlantard 
No, but the format function is the same.

Recently we only add CLI support and some code refactoring.
Will update when the function different.

And after the CLI become more mature maybe we can integrate into brew.
Thanks for @saagarjha 

BTW I'll close this issue since XCode 8 is supported now.
